### PR TITLE
Implement Jörmungandr Transaction-message decoder

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#218 

# Overview

- Implemented `Transaction` message
- Tested through the existing single unit-test

# Comments

- The second commit is for replacing `Discrimination` with `Network`

### CODE REVIEW

File: lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs [hunk 1]
```diff
@@ -141,14 +151,85 @@ getInitial = do
...
+        case tag of
+            1 -> isolate 128 $ do
+                -- Old address witness scheme
+                xpub <- getByteString 64
+                sig <- Hash <$> getByteString 64
+                return $ PublicKeyWitness xpub sig
+
+            2 -> isolate 64 $ do
+                _sig <- Hash <$> getByteString 64
+                error "unimplemented: New address witness scheme"
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:175]: Trying to implement as little as possible for now…

File: lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs [hunk 2]
```diff
@@ -70,8 +80,9 @@ data TODO = TODO
 data SignedVote = SignedVote
     deriving (Eq, Show)
 
+-- Do-notation is favoured over applicative syntax for readability:
+{-# ANN module ("HLint: ignore Use <$>" :: String) #-}
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:84]: :+1:

File: lib/jormungandr/test/unit/Cardano/Wallet/Binary/JormungandrSpec.hs [hunk 3]
```diff
@@ -72,7 +73,15 @@ genesisBlock = Block genesisHeader
         , AllowAccountCreation True
         , ConfigLinearFee (LinearFee 0 0 0)
         ]
-    , UnimplementedMessage 2
+    , Transaction $ Tx
+        { inputs = []
+        , outputs =
+            [ TxOut
+                { address = Address "3$\195xi\193\"h\154\&5\145}\245:O\"\148\163\165/h^\ENQ\245\248\229;\135\231\234E/"
+                , coin = Coin 14
```
Comment (KtorZ) [inline lib/jormungandr/test/unit/Cardano/Wallet/Binary/JormungandrSpec.hs:81]: `Coin 14` sounds like something I've could have put in the genesis block indeed :joy:

**Review by KtorZ — APPROVED** [review]
LGTM.

File: lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs [hunk 4]
```diff
@@ -141,14 +152,85 @@ getInitial = do
...
+            other -> fail $ "Invalid witness type: " ++ show other
+
+
+{-------------------------------------------------------------------------------
+                            Common Structure
+-------------------------------------------------------------------------------}
+
+getTokenTransfer :: Get ([TxIn], [TxOut])
+getTokenTransfer = do
+    inCount <- fromIntegral <$> getWord8
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:188]: don't we miss transaction header (2bytes) here according to https://github.com/input-output-hk/rust-cardano/blob/master/chain-impl-mockchain/doc/format.md#token-transfer

File: lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs [hunk 5]
```diff
@@ -141,14 +152,85 @@ getInitial = do
...
+        value <- Coin <$> getWord64be
+        return $ TxOut addr value
+
+    getAddress = do
+        headerByte <- getWord8
+        let kind = kindValue headerByte
+        let _discrimination = discriminationValue headerByte
+        case kind of
+            -- Single Address
+            0x3 -> Address <$> getByteString 32
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:212]: where can I find this info? I can't seem to find it in https://github.com/input-output-hk/rust-cardano/blob/master/chain-impl-mockchain/doc/format.md#preliminaries

**Review by akegalj — COMMENTED** [review]
there are two points that are confusing to me - maybe I am missing something

File: lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs (hunk 4 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:188]: It works this way 🤷‍♂ (+ I guess I followed the rust-decoder)

File: lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs (hunk 5 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:212]: Should be from here https://github.com/input-output-hk/rust-cardano/blob/e0616f13bebd6b908320bddb1c1502dea0d3305a/chain-addr/src/lib.rs#L392-L443

**Pull request merged**